### PR TITLE
nextcloud: update to 25.0.2

### LINKED ISSUE #2211: Upgrade to Nextcloud 25
Nextcloud released v25 a few days ago. The snap is a bit behind currently.

### PR DESCRIPTION
**Pull request by Pilzinsel64**
Fix #2211 
Fix #2222 

Don't want to attacked you @stondino00, just had the time for yet. Feel free to close this PR if not needed.

Anyways, not sure, if the tests success, as mentoined in [this comment](https://github.com/nextcloud-snap/nextcloud-snap/pull/2222#issuecomment-1302154488). No clue how fix them myself right now.

Also, not sure if it's right to change the changelog already here. Not sure what your process is for. Just a suggestion from my side.

### CODE REVIEW

**Comment by kyrofa** [comment]
> Don't want to attacked you @stondino00, just had the time for yet. Feel free to close this PR if not needed.

No apology needed, we're always grateful for help! There's no rule saying that @stondino00 needs to do this, he just tends to be the fastest :) . Thanks for the PR!

File: CHANGELOG [hunk 1]
```diff
@@ -1,3 +1,6 @@
+v 25.0.2snap1
```
Comment (kyrofa) [inline CHANGELOG:1]: Please remove this, we generate the changelog when creating a release as opposed to on the fly.

**Review by kyrofa — CHANGES_REQUESTED** [review]

**Comment by stondino00** [comment]
Good work @Pilzinsel64

**Comment by Pilzinsel64** [comment]
Changes made. Hope that's good now.

**Comment by Pilzinsel64** [comment]
Regarding the error at the integration test:
```
Logging in Logging in with correct credentials
     Failure/Error: fill_in "User", with: "admin"

     Capybara::ElementNotFound:
       Unable to find field "User" that is not disabled
     # ./vendor/bundle/ruby/2.6.0/gems/capybara-3.35.3/lib/capybara/node/finders.rb:303:in `block in synced_resolve'
     # ./vendor/bundle/ruby/2.6.0/gems/capybara-3.35.3/lib/capybara/node/base.rb:83:in `synchronize'
     # ./vendor/bundle/ruby/2.6.0/gems/capybara-3.35.3/lib/capybara/node/finders.rb:292:in `synced_resolve'
     # ./vendor/bundle/ruby/2.6.0/gems/capybara-3.35.3/lib/capybara/node/finders.rb:53:in `find'
     # ./vendor/bundle/ruby/2.6.0/gems/capybara-3.35.3/lib/capybara/node/actions.rb:91:in `fill_in'
     # ./vendor/bundle/ruby/2.6.0/gems/capybara-3.35.3/lib/capybara/session.rb:768:in `block (2 levels) in <class:Session>'
     # ./vendor/bundle/ruby/2.6.0/gems/capybara-3.35.3/lib/capybara/dsl.rb:58:in `block (2 levels) in <module:DSL>'
     # ./integration/login_spec.rb:4:in `block (2 levels) in <top (required)>'
```

This seems to be the part of the definition. But as I said [here](https://github.com/nextcloud-snap/nextcloud-snap/pull/2222#issuecomment-1302154488), the user field is here with the same name, as well as the password field.
https://github.com/Pilzinsel64/nextcloud-snap/blob/fade25fabff9f32ad6f2e8709d9ed95b2dbe4c9a/tests/integration/login_spec.rb#L3-L5

Would it be an idea to rename them to lower-case? `User`->`user` and `Password` to `password`? Might be random, but the only idea from my side.

File: CHANGELOG (hunk 1 above)
Comment (Pilzinsel64) [reply CHANGELOG]: Done.

**Comment by kyrofa** [comment]
I'm not sure offhand what's going on with the integration tests @Pilzinsel64. Are you able to build the snap locally and run them yourself? I think that's what's going to be required here. I'll try to get to it ASAP.

**Comment by Pilzinsel64** [comment]
Well, I need a little bit to get into [this instructions](https://github.com/nextcloud-snap/nextcloud-snap/wiki/Building-your-first-Nextcloud-snap) and setup a new VM. But I sould be able to get it working within the next 24 hours.
I keep you updated.

**Comment by Pilzinsel64** [comment]
So, I treid setting up on several VMs, on a Raspberry Pi 4... I failed. I have not the hardware to test it probably and I can't get VMs to work with multipass, whatever hack I use for enabling.
grafik

Sorry, but I can't continue on this today and not sure if I can tomorrow. Maybe you have a tip for me how to get multipass working.

**Comment by Pilzinsel64** [comment]
@kyrofa Alternatively, is it possible for you to provide this build to a public channel for this PR like you did for a few other PRs? The snap seems to build fine and also installs. So it's more a question of, does the website even run, and testing a packaged sap is absolutely possible for me on a test VM.

**Comment by Pilzinsel64** [comment]
Finally got it a step fordward with lxd after another four hours...

Now I stuck at pulling PHP extensions. @kyrofa Anything I can do here? Without the `--debug` flag it told me that "snapcraft" has no attribute "sources".
grafik

**Comment by Pilzinsel64** [comment]
Finally, after another six hours and observing the integration test log, I got the snap to build!
I've installed the snap and set up nextcloud, clicked a bit around and here we go, it works!
grafik
grafik

Also, there are no unexcpected errors and warnings here! :)
grafik

Runs very fast and smooth (always forget how fast NC could be...), as far as I can tell. But I'm not sure If I want to upgrade my productive instance to such a custom build only for testing. But I will do as soon as a test build is at like latest/beta/pr-2247 or latest/beta.

**Comment by petre2dor** [comment]
> https://github.com/nextcloud-snap/nextcloud-snap/blob/362d5cf2d4052cc0681ec75d121c8b5ba12fc259/tests/integration/login_spec.rb#L3-L5
Is the `visit /` following the redirect to `/login`?

**Comment by Pilzinsel64** [comment]
> Is the `visit "/"` following the redirect to `/login` or to `/index.php/login`?

Following to `/index.php/login` as pretty urls are not enabled and not possible.

**Comment by Pilzinsel64** [comment]
@kyrofa I know you are busy, but have two questions yet (sorry).

Is there *anything* we can do here for you?
I had a few tests with the [login_spec.rb](https://github.com/Pilzinsel64/nextcloud-snap/blob/nc2502/tests/integration/login_spec.rb) file, but without luck, as I have absolutely no knowledge with testing frameworks like Capybara. And [here](https://www.rubydoc.info/gems/capybara/Capybara%2FNode%2FActions:fill_in) are the specifications I found for the `fill_in` command.

Can you you provide us a beta build at the channel `latest/beta/pr-2247`, please, so we can do a little bit testing over christmas as you have time enough to have a look to the integration test? Or is that too much work to push such a build? Don't know if you push builds for a PR manually or if that goes mostly automatically too.

**Comment by kyrofa** [comment]
> Don't know if you push builds for a PR manually or if that goes mostly automatically too.

It used to be automatic, but Canonical keeps breaking stuff, so it's manual these days. I just started the build, I'll let you know when it's ready.

**Comment by kyrofa** [comment]
Alright, builds of this snap for all architectures is available on `latest/beta/pr-2247`. Also, I had some time to play with the tests, please see #2254.

**Comment by Pilzinsel64** [comment]
Hm, updating from v24.0.7 says a downgrade is not possible. 🤔 
```
error: cannot perform the following tasks:
- Run post-refresh hook of "nextcloud" snap if present (run hook "post-refresh": Nextcloud doesn't support downgrades, but you can revert. See 'snap revert -h' for details.)
```

**Comment by kyrofa** [comment]
Whoops, I missed a step when releasing 24.0.7 a while back. I'll have to re-spin these builds, I'm sorry! @Pilzinsel64 this branch's history is a little yucky at this point, would you mind rebasing onto develop and squashing it down?

**Comment by Pilzinsel64** [comment]
Uh, rebase seems to be not possible as the branche is up-to-date with upstream. And squashing seems to be not possible too (`Unable to squash. Squashing replays all commits up to the last one required for the squash. A merge commit cannot exist among those commits.`).

**Comment by Pilzinsel64** [comment]
Wow... Just found out how to force rebase. Now we broke everything I guess.

**Comment by kyrofa** [comment]
Haha, nah don't worry, it looks good now.

**Comment by Pilzinsel64** [comment]
Oh, true. Wow. Let me remove my clone on my hard drive then. You won't see what I probably would have pushed If I hadn't a look to this page again before. :D

**Comment by kyrofa** [comment]
Git itself does this kind of thing for you. In this case, `git rebase -i develop` and then change all but the first `pick`s to `squash`.

**Comment by kyrofa** [comment]
Alright, still not quite there yet, it seems. Progress, at least! This test is failing because the maintenance mode screen is still up when it tries to run that check. I'll take another pass in a bit.

**Comment by kyrofa** [comment]
Alright, newly-built snaps for all architectures are available in `latest/beta/pr-2247`.

**Comment by Pilzinsel64** [comment]
I've updated my private instance just now. The upgrade process run smoothly. Seems good now!

What I did see, but not related to the snap itself:
- The only problem I had is a bug in core Nextcloud, that every Design configuration has been lost and an admin need to re-configure them and disable costimization for all users (otherwise they will never see your global custom color). That's a bug in core Nextcloud that I already had at two other instances. [Look here.](https://github.com/nextcloud/server/issues/34578#issuecomment-1286055991)
- The app "Polls" updates very very long.
- The app "Bookmarks" need to be re-configured a little bit, as they changes something about the preview generation.
- The app "Custom menu" wont update for some reason. Need to investiage a bit later this day.

If I get anything else, related or unrelated to the snap package, I will let you know here.

**Comment by kyrofa** [comment]
Historically we haven't really cared about issues for non-default apps, but tell me more about this one @Pilzinsel64:

> The only problem I had is a bug in core Nextcloud, that every Design configuration has been lost and an admin need to re-configure them and disable costimization for all users (otherwise they will never see your global custom color). That's a bug in core Nextcloud that I already had at two other instances. https://github.com/nextcloud/server/issues/34578#issuecomment-1286055991

@szaimen's comment leads me to believe that should have been solved in this release. It sounds like perhaps it is not?

**Comment by Pilzinsel64** [comment]
Sure, the points above I left just for other people that maybe get into similar issues.

@kyrofa Not sure if my issue(s) are real bugs but this happens after the upgrade to NC25 (my private instance on the pic, but it's the same on my other two instances):
grafik

Sadly I didn't take pictures of directly after the upgrade.

I don't see it as critical blocker imao as it will not break anything but you just need to take care of this configurations. It's probably just me having that issues at three instances with different setup, not sure why. Or if it isn't intended, then it probably not getting fixed (soon).

**Review by kyrofa — APPROVED** [review]
Alright, this seems to work well both as a fresh install and as an upgrade. It sounds like there are some issues with theming, but there are always issues with theming and we disable it by default as a result anyway. I think it's good enough to ship. Thanks for the hard work, @Pilzinsel64!

**Pull request merged**